File: php/src/Attribute.php
<?php

declare(strict_types=1);

namespace HtmlSanitization;

use FFI;

/**
 * A DOM attribute, **borrowed** for the duration of a callback.
 *
 * Do not retain one past the callback that gave it to you — the DOM is freed
 * when {@see HtmlSanitizer::sanitize()} returns.
 */
final class Attribute
{
    /** @param FFI\CData $ptr */
    public function __construct(private FFI $ffi, private $ptr)
    {
    }

    public function name(): string
    {
        return Native::takeString($this->ffi, $this->ffi->aether_hs_embed_attr_name($this->ptr));
    }

    public function value(): string
    {
        return Native::takeString($this->ffi, $this->ffi->aether_hs_embed_attr_value($this->ptr));
    }

    /**
     * Rewrite the value in place (e.g. to canonicalise a URL rather than
     * remove the attribute). The engine copies the string, so PHP's transient
     * buffer is safe here.
     */
    public function setValue(string $value): void
    {
        $this->ffi->aether_hs_embed_attr_set_value($this->ptr, $value);
    }

    public function __toString(): string
    {
        return sprintf('Attribute(%s=%s)', $this->name(), $this->value());
    }
}

File: java/src/main/java/org/htmlsanitizer/Attribute.java
package org.htmlsanitizer;

import java.lang.foreign.Arena;   // referenced from setValue's javadoc
import java.lang.foreign.MemorySegment;

/**
 * A DOM attribute, <b>borrowed</b> for the duration of a callback.
 *
 * <p>Do not retain one past the callback that gave it to you — the DOM is
 * freed when {@code sanitize} returns.
 */
public final class Attribute {

    private final Native api;
    private final MemorySegment ptr;

    Attribute(Native api, MemorySegment ptr) {
        this.api = api;
        this.ptr = ptr;
    }

    public String name() {
        try {
            return api.takeString((MemorySegment) api.attrName.invokeExact(ptr));
        } catch (Throwable t) {
            throw Native.wrap(t);
        }
    }

    public String value() {
        try {
            return api.takeString((MemorySegment) api.attrValue.invokeExact(ptr));
        } catch (Throwable t) {
            throw Native.wrap(t);
        }
    }

    /**
     * Rewrite the attribute's value in place — e.g. to canonicalise a URL
     * rather than let the attribute be removed.
     *
     * <p>The ABI stores the pointer we pass straight into the DOM node
     * ({@code at.value = value} in {@code core/embed.ae}) rather than copying
     * it, so the buffer must outlive the callback. A confined {@link Arena}
     * would be freed at the end of this method and leave the DOM pointing at
     * released memory; a libc-{@code malloc}'d copy stays valid for the rest
     * of the sanitize run. It is not freed here — the sanitizer core owns it once
     * stored, the same contract as {@code on_filter_url}'s return value.
     */
    public void setValue(String v) {
        try {
            MemorySegment s = api.mallocString(v);
            api.attrSetValue.invokeExact(ptr, s);
        } catch (Throwable t) {
            throw Native.wrap(t);
        }
    }

    @Override
    public String toString() {
        return "Attribute(" + name() + "='" + value() + "')";
    }
}
